### ISSUE TITLE
[clang-tidy] `bugprone-misplaced-widening-cast` shouldn't warn on correct `constexpr` values

### ISSUE THREAD

**Issue by Jacobfaib** [llvm/llvm-project#208078]
Consider the following (https://godbolt.org/z/5jTTWh9cb):
```c++
#include <cstddef>

void bar(std::size_t);

void foo() {
  constexpr int x = 256;

  bar(static_cast<std::size_t>(x * 2));
}
```
The `static_cast<std::size_t>` is indeed pointless if the intent was to avoid overflow of `int` before passing to `bar()`. The correct form would be
```c++
static_cast<std::size_t>(x) * 2
```
However in this case, because `x` is `constexpr`, the compiler (and by extension `clang-tidy`) can compute the value of `x * 2` and see that it never overflows. Even if the cast is pointless, the code is still provably correct as written.

In this case `clang-tidy` should suppress the `misplaced-widening-cast` diagnostic, because otherwise it is pointless churn. There is no "cast" occurring anyways, the compiler will just constant fold the value to `size_t` anyways.

**Comment by Jacobfaib** [comment]
To elaborate a bit more, `clang-tidy` should continue to emit warnings for:

1. Runtime values, i.e. if `constexpr int x` becomes `const int x`. Even if `x` is still technically constant-foldable, `constexpr` is deliberately forcing constant evaluation, so `const` should signal deliberate runtime-ish evaluation.
2. Overflowing casts detected at compile-time. I.e.
```c++
constexpr int x = -1;

static_cast<std::size_t>(x * 2); // oops, cast might not do what you think it does
```

**Comment by vbvictor** [comment]
Can you just remove the cast in first example: https://godbolt.org/z/adT35874q?
No cast -> no warning

**Comment by vbvictor** [comment]
I don't think we should add special handling of constexpr case. If you know that it doesn't overflow - why put cast in the first place.
Unless I'm missing something here.

**Comment by Jacobfaib** [comment]
Yes, there are many ways to silence/address the warning, but they are ultimately churn. The code is already provably correct, and there is no "bug" to be fixed here from the compilers perspective.

I guess my point is this, this warning is supposed to catch instances where either:

1. The cast is not doing what the author thinks it's doing, or
2. The author _has_ considered the case carefully but still misplaced the cast or cast to the wrong thing

But here _it does not matter_. Even if the authors mental model of the cast is wrong, the code is still unambiguously correct. Even if I go and change the cast it still produces the same program, byte-for-byte for all compilers on the planet.

Perhaps a middle ground would be to introduce a pedantic mode where `clang-tidy` emits warnings even in these cases?

> If you know that it doesn't overflow - why put cast in the first place.

Some compilers will now warn about signed/unsigned conversion. Hence the cast.

**Comment by Jacobfaib** [comment]
> I don't think we should add special handling of constexpr case. 

I might argue that there is some precedence for special-casing `constexpr` already :). Consider the following:
```c++
constexpr int x = 256;
unsigned y{x};
```
There is no diagnostic for the above. `clang` _could_ emit one, because on its face there is a possibility for narrowing, but `x` is `constexpr` and therefore `y` is computable and `clang` can prove the conversion does not narrow. But for any of the following cases:
```c++
// 1.
constexpr int x = -1;
unsigned y{x}; // error

// 2.
int x = 256;
unsigned y{x}; // error
```
`clang` emits a diagnostic because they might narrow (or I guess in case 1, does). My request is simply that this extends to the `clang-tidy` checks as well.

**Comment by shivaansharma** [comment]
can this be assigned to me? I would love to work on it .
